pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,889 @@
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			chClient:   mockAdapter,
+			syncKeeper: make(map[string]chan model.Message),
+			stopChan:   make(chan struct{}),
+		}, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ehc := &EdgeHubController{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.receiveErr {
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage(""), errors.New("Connection Refused")).AnyTimes()
+				go ehc.routeToEdge()
+				stop := <-ehc.stopChan
+				if stop != struct{}{} {
+					t.Errorf("TestRouteToEdge error got: %v want: %v", stop, struct{}{})
+				}
+			} else {
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.EdgedGroup, "", ""), nil).Times(1)
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.TwinGroup, "", ""), nil).Times(1)
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage(""), errors.New("Connection Refused")).Times(1)
+				go ehc.routeToEdge()
+				stop := <-ehc.stopChan
+				if stop != struct{}{} {
+					t.Errorf("TestRouteToEdge error got: %v want: %v", stop, struct{}{})
+				}
+			}
+		})
+	}
+}
+
+//TestSendToCloud() tests whether the send to cloud functionality works properly
+func TestSendToCloud(t *testing.T) {
+	initMocks(t)
+	msg := model.NewMessage("").BuildHeader("test_id", "", 1)
+	msg.Header.Sync = true
+	tests := []struct {
+		name    string
+		fields  EdgeHubController
+		message model.Message
+		wantErr bool
+		waitErr bool
+	}{
+		{"TestSendToCloud: Positive Test Case", EdgeHubController{
+			context:  context.GetContext(context.MsgCtxTypeChannel),
+			chClient: mockAdapter,
+			config: &config.ControllerConfig{
+				HeartbeatPeroid: 10 * time.Second,
+			},
+			syncKeeper: make(map[string]chan model.Message),
+		}, *msg, false, false},
+
+		{"TestSendToCloud: Wait Error", EdgeHubController{
+			chClient: mockAdapter,
+			config: &config.ControllerConfig{
+				HeartbeatPeroid: 10 * time.Second,
+			},
+			syncKeeper: make(map[string]chan model.Message),
+		}, *msg, false, true},
+
+		{"TestSendToCloud:Send Failure", EdgeHubController{
+			chClient:   mockAdapter,
+			syncKeeper: make(map[string]chan model.Message),
+		}, model.Message{}, true, false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ehc := &EdgeHubController{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.wantErr {
+				mockAdapter.EXPECT().Send(gomock.Any()).Return(errors.New("Connection Refused")).Times(1)
```

Comment:
instead of hardcoding this, can we do something we have done in this file
https://github.com/kubeedge/kubeedge/blob/master/pkg/devicetwin/dtclient/device_db_test.go#L380
This might help in avoiding if-else conditions in the code ?